include/rcpputils/clamp.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 PAL Robotics S.L.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * \file clamp.hpp
+ * \brief Restrict a value between two bounds.
+ */
+
+#ifndef RCPPUTILS__CLAMP_HPP_
+#define RCPPUTILS__CLAMP_HPP_
+
+#include <cassert>
+
+namespace rcpputils
+{
+
+/**
+ * If v compares less than lo, returns lo; otherwise if hi compares less
+ * than v, returns hi; otherwise returns v. Uses operator< to compare the values.
+ *
+ * \param[in] v The value to clamp.
+ * \param[in] lo The lower boundary.
+ * \param[in] hi The higher boundary.
+ * \return Reference to lo if v is less than lo, reference to hi if hi is less than v, otherwise
+ * reference to v.
+ * \note Implementation from https://en.cppreference.com/w/cpp/algorithm/clamp.
+ * \warning Capturing the result of clamp by reference if one of the parameters is rvalue produces
+ *  a dangling reference if that parameter is returned.
+ */
+template<class T>
+[[deprecated("use std::clamp instead, introduced in C++17")]]
```

Comment:
It's kind of odd to be adding a function that is immediately deprecated.

If the goal here is to completely rely on `std::clamp` instead of this version, then I'll suggest that we do *not* move this header out of `rcppmath` for now, but instead just mark it deprecated in there.  Any downstream users will have to change from `rcppmath::clamp` to `std::clamp` directly, and then in J-Turtle we can just remove everything under the `rcppmath` namespace.

Reply:
But does it hurt to move it and deprecate it in favor of `std::clamp`? I would just like to completely gut that namespace. I could also see that we "tock" these two things separately. I.e. we could drop the `rcppmath` namespace in the next release, but for some reason not move forward with the use of `std::clamp`.

If you really don't want it as-is, I would prefer to just move the namespace and leave the `std::clamp` thing for later. I only did that here because I noticed as I was moving the functions. The primary goal here is to deprecate the `rcppmath` namespace.

Reply:
Ok, I still don't really see how undoing changes in this pr makes it better, but I went a head and did it. I had to rebase and stuff, so everything is in [7ec674f](https://github.com/ros2/rcpputils/pull/165/commits/7ec674fec97965128ee0e79e6b32b7486affdcf4) now, including `uncrustify` fixes.

Reply:
Just to recap, I think I've updated it as you suggested. I think that it's fine in this state, so if you agree we can merge it as-is (with CI)?